test/soak/soak_test.go

```diff
@@ -0,0 +1,142 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: MIT
+package soak
+
+import (
+	"fmt"
+	"log"
+	"runtime"
+	"strings"
+	"testing"
+
+	"github.com/aws/amazon-cloudwatch-agent-test/internal/common"
+	"github.com/shirou/gopsutil/v3/process"
+	"github.com/stretchr/testify/require"
+)
+
+func TestSoakLow(t *testing.T) {
+	switch runtime.GOOS {
+	case "darwin":
+		// todo:
+	case "linux":
+		runTest(t, "SoakTestLowLinux", "resources/soak_linux.json", 5, 100_000_000)
+	case "windows":
+		runTest(t, "SoakTestLowLinux", "resources/soak_windows.json", 5, 100_000_000)
+	}
+}
+
+// todo: add high througput
+// todo: logrotate
+// todo: multiple-logs
+func TestSoakHigh(t *testing.T) {
+	switch runtime.GOOS {
+	case "darwin":
+		// todo:
+	case "linux":
+		runTest(t, "SoakTestHighLinux", "resources/soak_linux.json", 50, 200_000_000)
+	case "windows":
+		runTest(t, "SoakTestHighWindows", "resources/soak_windows.json", 50, 200_000_000)
+	}
+}
+
+// runTest just does setup.
+// It starts the agent and starts some background processes which generate
+// load and monitor for resource leaks.
+// The agent config should use a mocked backend (local stack)to save cost.
+// testName is used as the namespace for validator metrics.
+func runTest(t *testing.T, testName string, configPath string, cpuLimit int, memLimit int) {
+	require.NoError(t, startLocalStack())
+	common.CopyFile(configPath, common.ConfigOutputPath)
+	require.NoError(t, common.StartAgent(common.ConfigOutputPath, false))
+	require.NoError(t, startValidator(testName, cpuLimit, memLimit))
+	if strings.Contains(testName, "HighLoad") {
+		require.NoError(t, startLogGen(10, 1000, 100))
+		require.NoError(t, startEMFGen(10, 1000))
+		require.NoError(t, startStatsd(10, 1000, 1000))
+	} else {
+		require.NoError(t, startLogGen(10, 100, 100))
+		require.NoError(t, startEMFGen(10, 100))
+		require.NoError(t, startStatsd(1, 100, 100))
+	}
+}
+
+// Refer to https://github.com/localstack/localstack
+func startLocalStack() error {
```

Comment:
We will need docker installed on whatever Windows AMI we test against.
It shouldn't be an issue though since we already use custom AMIs for patching purposes.